### PR TITLE
Create new donor upload file

### PR DESCRIPTION
**Pull request by arvindraghavan**
Changes: 
- Refactor CSV creation function
- Add logic for generating new donor record

Review:
- [ ] Kevin

### CODE REVIEW

**Comment by arvindraghavan** [comment]
@kdawgsf, I have only a couple of cases that now result in invalid records, which I will fix shortly. Please see this latest commit if you reviewing

File: update_existing_families.py [hunk 1]
```diff
@@ -149,8 +160,68 @@ def load_csv_file(filename, expected_headers):
...
-        print("TODO Create record for new donor / new student %s and add to dp_import_studentrecords" % (student_id))
+        # At this point we know that: 
+        #   (a) this student isn't in DP (i.e., this is new student)
+        #   (b) this student doesn't have an sibling currently in BSD
+        # Chances are the donor is also new, unless 
+        #   (i) the donor has a kid that has graduated already OR, 
+        #   (ii) this student is returning after a break from BSD, but somehow was given a new student ID
+        # If either (i) or (ii) is true, then DP will not create a new record, but update the existing donor record, so we should be okay. 
+        # At any rate, we have to prepare a new record for this donor, with several custom fields
+        print("Creating record for new donor w/ new student %s" % (student_id))
```
Comment (kdawgsf) [inline update_existing_families.py]: I'd rather track some counts for stuff like this and emit counts at the end. The specific data is available in the csv file.

Comment (arvindraghavan) [inline update_existing_families.py]: I was thinking exactly the same thing. We should only emit some succinct stats to stdout

Comment (arvindraghavan) [inline update_existing_families.py]: Will fix this in a couple of mins.

File: update_existing_families.py [hunk 2]
```diff
@@ -149,8 +160,68 @@ def load_csv_file(filename, expected_headers):
...
+                'ADDRESS_TYPE': 'HOME',
+                'EMAIL': email,
+                'SPOUSE_EMAIL': district_record['Parent2Email'],
+                'HOME_PHONE': district_record['home_phone'],
+                'MOBILE_PHONE': district_record['Parent1DayPhone'],
+                'SPOUSE_MOBILE': district_record['Parent2DayPhone'],
+                'STU_FNAME': district_record['Student First Name'],
+                'STU_LNAME': district_record['Student Last Name'],
+                'STU_NUMBER': district_record['SystemID'],
+                'SCHOOL': district_record['School'],
```
Comment (kdawgsf) [inline update_existing_families.py]: Need to convert this to the DP school code

File: update_existing_families.py [hunk 3]
```diff
@@ -149,8 +160,68 @@ def load_csv_file(filename, expected_headers):
...
+            email = district_record['Parent1Email']
+            
+            
+        newdonorrecord = {
+                'FIRST_NAME': h_f_name,
+                'LAST_NAME': h_l_name,
+                'SP_FNAME': w_f_name,
+                'SP_LNAME': w_l_name,
+                'SALUTATION': salutation,
+                'INFORMAL_SAL': h_f_name + " " + w_f_name,
```
Comment (kdawgsf) [inline update_existing_families.py]: You're missing "and". Also, you're not handling the case where the district data only has 1 parent.

File: update_existing_families.py [hunk 4]
```diff
@@ -149,8 +160,68 @@ def load_csv_file(filename, expected_headers):
...
+        print("Creating record for new donor w/ new student %s" % (student_id))
+
+        h_f_name = district_record['Parent 1 First Name']
+        w_f_name = district_record['Parent 2 First Name']
+        h_l_name = district_record['Parent1 Last Name']
+        w_l_name = district_record['Parent 2 Last Name']
+        if w_l_name == h_l_name :
+            salutation = h_f_name + " and " + w_f_name + " " + w_l_name
+        else:
+            salutation = h_f_name + " " + h_l_name + " and " + w_f_name + " " + w_l_name
```
Comment (kdawgsf) [inline update_existing_families.py]: Need to handle the case where the district data only has 1 parent.

File: update_existing_families.py [hunk 5]
```diff
@@ -149,8 +160,68 @@ def load_csv_file(filename, expected_headers):
...
+            email = district_record['Parent2Email']
+        else:
+            email = district_record['Parent1Email']
+            
+            
+        newdonorrecord = {
+                'FIRST_NAME': h_f_name,
+                'LAST_NAME': h_l_name,
+                'SP_FNAME': w_f_name,
+                'SP_LNAME': w_l_name,
```
Comment (kdawgsf) [inline update_existing_families.py]: If there is only one parent in the district data, and that parent is female, the parent seems to go as parent 2. In DP in this case, the single parent goes as first/last and there is no spouse.

File: update_existing_families.py [hunk 6]
```diff
@@ -149,8 +160,68 @@ def load_csv_file(filename, expected_headers):
...
+                'SALUTATION': salutation,
+                'INFORMAL_SAL': h_f_name + " " + w_f_name,
+                'OPT_LINE': w_f_name + " " + w_l_name,
+                'ADDRESS': district_record['street'],
+                'CITY': district_record['street'],
+                'STATE': district_record['street'],
```
Comment (kdawgsf) [inline update_existing_families.py]: wrong field

File: update_existing_families.py [hunk 7]
```diff
@@ -149,8 +160,68 @@ def load_csv_file(filename, expected_headers):
...
+                'ZIP': district_record['street'],
```
Comment (kdawgsf) [inline update_existing_families.py]: wrong field

File: update_existing_families.py [hunk 8]
```diff
@@ -149,8 +160,68 @@ def load_csv_file(filename, expected_headers):
...
+            
+            
+        newdonorrecord = {
+                'FIRST_NAME': h_f_name,
+                'LAST_NAME': h_l_name,
+                'SP_FNAME': w_f_name,
+                'SP_LNAME': w_l_name,
+                'SALUTATION': salutation,
+                'INFORMAL_SAL': h_f_name + " " + w_f_name,
+                'OPT_LINE': w_f_name + " " + w_l_name,
```
Comment (kdawgsf) [inline update_existing_families.py]: If applicable

File: update_existing_families.py [hunk 9]
```diff
@@ -149,8 +160,68 @@ def load_csv_file(filename, expected_headers):
...
+                'ADDRESS': district_record['street'],
+                'CITY': district_record['street'],
```
Comment (kdawgsf) [inline update_existing_families.py]: wrong field

File: update_existing_families.py (hunk 5 above)
Comment (arvindraghavan) [inline update_existing_families.py]: This should be fixed too in the latest.

File: update_existing_families.py [hunk 10]
```diff
@@ -170,16 +170,30 @@ def save_as_csv_file(filename, header_fields, data):
...
+        if len(w_l_name) != 0:
+            if w_l_name == h_l_name :
+                salutation = h_f_name + " and " + w_f_name + " " + w_l_name
+            else:
+                salutation = h_f_name + " " + h_l_name + " and " + w_f_name + " " + w_l_name
+            informal_sal = h_f_name + " and " + w_f_name
+        else:
+            salutation = h_f_name + " " + h_l_name 
+            informal_sal = h_f_name 
+
```
Comment (kdawgsf) [inline update_existing_families.py:196]: I think there are still bugs here. E.g. if there is only parent 2, you're not assigning them as the only parent. I think you want to break up 3 cases (only parent1, only parent2, both), and assign salutation, informal salutation, and both sets of first and last names in each case.

**Comment by kdawgsf** [comment]
LGTM

**Pull request merged**